### ISSUE TITLE
Start testcafe at BeforeAll stage rather than Before stage in cucumberjs

### ISSUE THREAD

**Issue by bhreinb** [DevExpress/testcafe#2882]
### Are you requesting a feature or reporting a bug?

Feature

### What is the current behavior?

Slowness running tests with testcafe integrated with cucumberjs 

### What is the expected behavior?

A faster experience especially when running against saucelabs or browserstack

### How would you reproduce the current behavior (if this is a bug)?

I can attach a sample application to help reproduce

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

https://github.com/
https://google.com/

### Specify your

* operating system: macOS High SIerra 10.13.6
* testcafe version: v0.22.0
* node.js version: v8.11.3

Hi there,

I have been able to integrate testcafe with cucumberjs. My edition is built on top of the work done in this repo

https://github.com/rquellh/testcafe-cucumber

I support the following integrations at this point in time

1. testing with all the major browsers 
2. testing in parallel
3. testing with docker
4. testing with saucelabs & browserstack

which is an evolution of the aforementioned repo. Unfortunately, where the solution suffers is speed (especially when I integrate with saucelabs). I detail the problem below

When you execute a plan in a cucumber runner you have the following cycles

BeforeAll (Happens prior to execution of test plan)
Before (Happens prior to execution of test scenario)

```
function runTest(browser) {
  testCafeServer = browser.includes('saucelabs:') ? testCafe() : testCafe('localhost');
  testCafeServer
    .then((tc) => {
      tcTestCafe = tc;
      tcRunner = tcTestCafe.createRunner();
      return tcRunner
        .src('./e2e/test_cafe/Script.js')
        .screenshots('./reports/screenshots/', true)
        .browsers(browser)
        .reporter('spec')
        .run();
    });
}

Before(async function setUp(testCase) {
  console.log(`Running Test Scenario '${testCase.pickle.name}' Under PID ${process.pid}`);
  runTest(this.setBrowser());
  return this.promisetestControllerHolder
    .then(testController => testController.maximizeWindow())
    .then(() => this.initialisePageObjectMap());
});
```

Ideally i'd like to move the runTest function into the BeforeAll stage and only setup the connection to the browser once in the test run, because doing it like it is now is slow, especially when integrated with SauceLabs or BrowserStack. However, moving it to BeforeAll introduces another issue described here

https://testcafe-discuss.devexpress.com/t/testcafe-and-cucumber-js/207/24

in that when running multiple scenarios, if one scenario failed it would fail the other scenarios with the same failure.

So the feature request is an ability to reset the test at the Before stage but not the browser connection, as the cucumber runner would have already captured the failed assertion from previous tests thus allowing the flow to continue naturally. That is probably the biggest integration issue with cucumberjs at this point in time. Thoughts?

Btw this ticket #2835 can be closed.

**Comment by AndreyBelym** [comment]
Hi @bhreinb! I think it's better to address this issue to the author of https://github.com/rquellh/testcafe-cucumber. It doesn't look like as a TestCafe's issue, but more like an issue of Cucumber integration.

Currently we don't have official cucumber integration. For now it's hard to implement proper integration with Cucumber, because TestCafe was designed as all-in-one tool (browser manipulation + test runner + test results reporting), but Cucumber need only the browser manipulation subsystem. Unfortunately, all TestCafe's subsystem are tightly coupled, so we need to spend a lot of time on decoupling and them and creating proper Cucumber integration based on stand-alone browser manipulation subsystem.

**Comment by bhreinb** [comment]
Hi @AndreyBelym 

Thanks for your response. Yeah it would be great if TestCafe just had the stand-alone browser manipulation subsystem. That would be useful to a-lot of users I would think.

I'm fairly sure the author of that repo can't do anything to fix this. Functionality it works per design when I have the TestCafe setup in the Before stage but way too slow. In the BeforeAll (where I should be doing this but can't because of how TestCafe is designed)...because one failure causes all other tests to fail with the same exception..hence why I logged it to this repo hoping that you would give me ideas on how to reset a failed assertion.
